Use new slash commands API

jedenastka opened the pull request:
Switches to the new slash commands API, as the old one is no longer functional.

<img src="https://github.com/Aliucord/Aliucord/assets/33988779/593f35e7-721e-4a2e-968b-5b8fac09edc5" width="300px">

Fixes #412.

---

The `GUILD_APPLICATION_COMMAND_INDEX_UPDATE` event is currently not handled, so if an application updates it's commands the client needs to be restarted to see the changes.

I plan to implement this, as well as other improvements (see the unresolved `TODO`s in the code), as a separate MR, so basic functionality is brought to users sooner, but if it's necessary I can do it as part of this one.

There's also something that looks like a bug, but I'm unsure if it actually is one, as I don't know how the commands worked when the old API was functional - the command list does not come up immediately after typing `/`, it's required to tap on one of the bot icons to see it:

<img src="https://github.com/Aliucord/Aliucord/assets/33988779/e23d916e-159e-4138-8285-bd5d4bdb9d09" width="300px">

I will try to research this more, but when I asked about this on Discord, Juby210 said that's how he remembers it working.

jedenastka commented:
Current state of the MR:

- A crash when loading a guild with command permissions set up was discovered and fixed.
- The command list not coming up immediately was confirmed to be a bug and fixed.
- Most of the code improvements I planned were implemented as well due to the refactoring required to fix the other issues.
- The `GUILD_APPLICATION_COMMAND_INDEX_UPDATE` event is still not handled.

rushiiMachine commented:
> A crash when loading a guild with command permissions set up was discovered and fixed.

The slash command model changed, now permissions are different, description is nullable, etc.
Model might just have to be replaced entirely

jedenastka commented:
- Applications are now sorted properly when browsing commands.
- Permissions are mostly implemented - guild administrators should be able to use all commands, but can't yet.
- The `GUILD_APPLICATION_COMMAND_INDEX_UPDATE` event is still not handled.

jedenastka commented:
- Permissions should be now fully implemented.
- A crash when checking configurations of permissions was discovered and fixed.
- A crash when trying to use commands in DMs with bots was discovered and fixed.

rushiiMachine commented:
is this ready for review?

jedenastka commented:
Not really, Dolfies found some new issues with it, so I wouldn't consider it ready.

jedenastka commented:
This was done in Java, and apparently Java is horrible. Feel free to reimplement this in Kotlin, just respect my copyright.

Apethesis commented:
Why was it closed and re-opened? 🤔

C1nderC0ated commented:
is this fix abandoned?

thennothinghappened commented:
> is this fix abandoned?

Well there's your answer :P

verticalsync commented:
> > is this fix abandoned?
> 
> Well there's your answer :P

Is there an approximate on when this pr will be finished

jedenastka commented:
> Is there an approximate on when this pr will be finished

It will be finished when it's done.

MrMasrozYTLIVE commented:
Is there a way to use it in its current state? Open PR repo and build aliucord from source?

master3395 commented:
I wish slash commands were working in aliucord, for now i have to switch back to discord app to use slash commands 😓

jedenastka commented:
This should be ready for review now.

mantikafasi commented on Aliucord/src/main/java/com/aliucord/coreplugins/slashcommandsfix/Patches.java:
```diff
@@ -0,0 +1,288 @@
+
+    @SuppressWarnings("unchecked")
+    public void loadPatches(Context context) throws Throwable {
+        this.handleGuildApplicationsUpdateMethod = StoreApplicationCommands.class.getDeclaredMethod("handleGuildApplicationsUpdate", List.class);
+        this.handleGuildApplicationsUpdateMethod.setAccessible(true);
+        this.handleDiscoverCommandsUpdateMethod = StoreApplicationCommands.class.getDeclaredMethod("handleDiscoverCommandsUpdate", List.class);
+        this.handleDiscoverCommandsUpdateMethod.setAccessible(true);
+        this.handleQueryCommandsUpdateMethod = StoreApplicationCommands.class.getDeclaredMethod("handleQueryCommandsUpdate", List.class);
+        this.handleQueryCommandsUpdateMethod.setAccessible(true);
+        this.applicationCommandCountField = Application.class.getDeclaredField("commandCount");
+        this.applicationCommandCountField.setAccessible(true);
```
cant you use `ReflectUtils.getMethodByArgs()` instead of this verbosity

jedenastka replied:
```
this.handleGuildApplicationsUpdateMethod = StoreApplicationCommands.class.getDeclaredMethod("handleGuildApplicationsUpdate", List.class);
this.handleGuildApplicationsUpdateMethod = ReflectUtils.getMethodByArgs(StoreApplicationCommands.class, "handleGuildApplicationsUpdate", new Object[]{List.class});
```
*This verbosity* is 26 characters less verbose than `ReflectUtils.getMethodByArgs()` lol

mantikafasi replied:
1: you should not need to create new Object[], you should be able to do just `this.handleGuildApplicationsUpdateMethod = ReflectUtils.getMethodByArgs(StoreApplicationCommands.class, "handleGuildApplicationsUpdate", List.class);`

2: you wont need to setAccessible(true) for every method, since getMethodByArgs already does it

jedenastka replied:
Ok, that does make it shorter.

jedenastka replied:
For some reason, applying these changes:

```diff
diff --git a/Aliucord/src/main/java/com/aliucord/coreplugins/slashcommandsfix/Patches.java b/Aliucord/src/main/java/com/aliucord/coreplugins/slashcommandsfix/Patches.java
index 64cfd2a..b42a68c 100644
--- a/Aliucord/src/main/java/com/aliucord/coreplugins/slashcommandsfix/Patches.java
+++ b/Aliucord/src/main/java/com/aliucord/coreplugins/slashcommandsfix/Patches.java
@@ -57,12 +57,9 @@ final class Patches {
 
     @SuppressWarnings("unchecked")
     public void loadPatches(Context context) throws Throwable {
-        this.handleGuildApplicationsUpdateMethod = StoreApplicationCommands.class.getDeclaredMethod("handleGuildApplicationsUpdate", List.class);
-        this.handleGuildApplicationsUpdateMethod.setAccessible(true);
-        this.handleDiscoverCommandsUpdateMethod = StoreApplicationCommands.class.getDeclaredMethod("handleDiscoverCommandsUpdate", List.class);
-        this.handleDiscoverCommandsUpdateMethod.setAccessible(true);
-        this.handleQueryCommandsUpdateMethod = StoreApplicationCommands.class.getDeclaredMethod("handleQueryCommandsUpdate", List.class);
-        this.handleQueryCommandsUpdateMethod.setAccessible(true);
+        this.handleGuildApplicationsUpdateMethod = ReflectUtils.getMethodByArgs(StoreApplicationCommands.class, "handleGuildApplicationsUpdate", List.class);
+        this.handleDiscoverCommandsUpdateMethod = ReflectUtils.getMethodByArgs(StoreApplicationCommands.class, "handleDiscoverCommandsUpdate", List.class);
+        this.handleQueryCommandsUpdateMethod = ReflectUtils.getMethodByArgs(StoreApplicationCommands.class, "handleQueryCommandsUpdate", List.class);
         this.applicationCommandCountField = Application.class.getDeclaredField("commandCount");
         this.applicationCommandCountField.setAccessible(true);
```

Causes a `NoSuchMethodException`:

```
[PluginManager] Failed to start core plugin SlashCommandsFix
java.lang.NoSuchMethodException: com.discord.stores.StoreApplicationCommands.handleGuildApplicationsUpdate [class java.lang.Class]
  at java.lang.Class.getMethod(Class.java, [PluginManager] Failed to start core plugin SlashCommandsFix:2072)
  at java.lang.Class.getDeclaredMethod(Class.java:2050)
  at com.aliucord.utils.ReflectUtils.getMethodByArgs(ReflectUtils.java:96)
  at com.aliucord.coreplugins.slashcommandsfix.Patches.loadPatches(Patches.java:60)
  at com.aliucord.coreplugins.slashcommandsfix.SlashCommandsFix.start(SlashCommandsFix.java:23)
  at com.aliucord.coreplugins.CorePlugins.startAll(CorePlugins.kt:56)
  at com.aliucord.Main.init(Main.java:282)
  at java.lang.reflect.Method.invoke(Native Method)
  at com.aliucord.injector.InjectorKt.init(Injector.kt:116)
  at com.aliucord.injector.InjectorKt.access$init(Injector.kt:1)
  at com.aliucord.injector.InjectorKt$init$1.beforeHookedMethod(Injector.kt:43)
  at de.robv.android.xposed.XposedBridge$HookInfo.callback(XposedBridge.java:352)
```

Even though the method definitely exists, and before the changes everything seems to work correctly. Not sure what I am doing wrong.

jedenastka replied:
As mentioned by Juby on Discord, this shouldn't be used here.

rushiiMachine replied:
> As mentioned by Juby on Discord, this shouldn't be used here.

What was this about again?

sudoshindo commented:
Need this
Waiting

jedenastka commented:
Just wanted to formally announce this should be once again ready for review. All (I think) new bugs found since the previous are now fixed.

> Is there a way to use it in its current state? Open PR repo and build aliucord from source?

Yes. See: [Message in #support](https://discord.com/channels/811255666990907402/811261298997460992/1210929772683599882) on Aliucord's Discord server.

rushiiMachine commented on Aliucord/src/main/java/com/aliucord/utils/RNSuperProperties.kt:
```diff
@@ -42,7 +42,7 @@ object RNSuperProperties {
     // update to latest Beta branch sometimes
-    const val versionCode = 183109
-    const val versionString = "183.9 - rn"
+    const val versionCode = 218111
```
I think these should probably be updated by now

rushiiMachine replied:
nvm

rushiiMachine commented on Aliucord/src/main/java/com/aliucord/coreplugins/slashcommandsfix/SlashCommandsFix.java:
```diff
@@ -0,0 +1,29 @@
+import com.aliucord.entities.CorePlugin;
+import de.robv.android.xposed.XposedBridge;
+
+public final class SlashCommandsFix extends CorePlugin {
```
Can you also override `isHidden`/`isRequired` as well and return true?

rushiiMachine commented on Aliucord/src/main/java/com/aliucord/coreplugins/slashcommandsfix/ApiApplication.java:
Could you move all the model classes to a sub-package? ie, `com/aliucord/coreplugins/slashcommandsfix/models` or something

rushiiMachine commented on Aliucord/src/main/java/com/aliucord/coreplugins/slashcommandsfix/Patches.java:
```diff
@@ -0,0 +1,315 @@
+                    return true;
+                }
+
+                var applicationId = remoteApplicationCommand.getApplicationId();
+                var isUser = this.requestApplicationIndex(new ApplicationIndexSourceUser())
+                    .applications
+                    .containsKey(applicationId);
+                if (isUser) {
+                    // Allow all user application commands
+                    return true;
+                }
```
Does this PR handle the guild permission that disables user apps in guilds?

rushiiMachine commented on Aliucord/src/main/java/com/aliucord/coreplugins/slashcommandsfix/Patches.java:
```diff
@@ -0,0 +1,315 @@
+        } else {
+            // Only create a DM index source for bots
+            var channel = storeChannelsSelected.getSelectedChannel();
+            var channelType = channel.D();
+            if (channelType == Channel.DM) {
+                var user = channel.z().get(0);
+                var userIsBot = Optional.ofNullable(user.e())
+                    .orElse(false);
+                if (userIsBot) {
+                    var channelId = channel.k();
+                    applicationIndexSource = Optional.of(new ApplicationIndexSourceDm(channelId));
+                }
```
Could you use the UserWrapper/ChannelWrapper classes here instead of directly calling the obfuscated methods
